cmd/gateway_controller/main.go

```diff
@@ -143,8 +216,35 @@ func main() {
 	}
 
 	setupLog.Info("starting manager")
+
 	if err := mgr.Start(ctx); err != nil {
-		setupLog.Error(err, "problem running manager")
+		setupLog.Error(err, "problem running controller manager")
 		os.Exit(1)
 	}
+
+}
+
+func main() {
+	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
+	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
+	flag.BoolVar(&enableLeaderElection, "leader-elect", false,
+		"Enable leader election for controller manager. "+
+			"Enabling this will ensure there is only one active controller manager.")
+	opts := zap.Options{
+		Development: true,
+	}
+	opts.BindFlags(flag.CommandLine)
+	flag.Parse()
+
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+	ctx := ctrl.SetupSignalHandler()
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go startAddonManager(ctx)
+	go startGatewayController(ctx)
+	wg.Wait()
+
+	<-ctx.Done()
```

Comment:
Would it be possible to add an addon manager runnable to the controller manager which is already managing multiple go routines for us:

```.go
if err := mgr.Add(ocm.AddonRunnable{}); err != nil {
	setupLog.Error(err, "unable to set up ocm addon manager")
	os.Exit(1)
}

type AddonRunnable struct{}

func (r AddonRunnable) Start(ctx context.Context) error {
	fmt.Println("starting add-on manager")
    ...

	if err := addonMgr.Start(ctx); err != nil {
		return err
	}

	<-ctx.Done()
	return nil
}
```

AFAIK, this should work, and will take care of what to do when the likes of leadership election is being used.
You can keep the ocm addon stuff in it's own package somewhere (pkg/ocm maybe), and import it into main to keep it a bit more organised, will also mean main.go is only changing slightly.

wdyt?

Reply:
Reworked. Now should be fine